### ISSUE TITLE
Hi,

### ISSUE THREAD

**Issue by sanjanasri** [dfdx/Boltzmann.jl#8]
I've been using your Boltzmann.jl for my project.  I'd like to first appreciate your amazing work. The question I've come across with is
a) How to find the error rate for each iteration. 
b) How can I get the unroll the dbn layers. and 
c) How can i change the functions inside rbm ( say, I wish to change tanh function instead of logistic ). 

Since I am newbie to Julia I dunno how to alter the source code. 
Any help would be appreciated.

**Comment by dfdx** [comment]
Hi, 

since RBMs and DBNs are probabilistic structures, they don't support error rate per se. Instead, concept of pseudo-likelihood is used to track learning progress. Normally, pseudo-likelihood is printed out during training, but you can also get it programmatically using [`score_samples`](https://github.com/dfdx/Boltzmann.jl/blob/master/src/rbm.jl#L110) function. E.g.:

```
using Boltzmann
rbm = ...
X = ...
fit(rbm, X, n_iter=10)
pseudo_likelihood = mean(Boltzmann.score_samples(rbm, X))
```
Note, that `score_samples()` is not exported by default, so you need to use qualified name, i.e. `Boltzmann.score_samples()`.

If you are primarily interested in DBNs, you need to decide what you want to measure. The easiest way is to take last RBM (e.g. `last_rbm = dbn.layers[end][2]`) and compute pseudo-likelihood for it. This is assuming you want to see how error changes after different number of iterations. 

Also note, that `fit()` is completely updatable, so you can, for example, run 10 iterations, check pseudo-likelihood and run 10 more iterations (so totaling in 20 iterations) to improve results a little bit more. 

I'm not sure what you mean by "unrolling" the network, but if you are looking for a way to fine-tune the network, take a look at [Mocha.jl](https://github.com/pluskid/Mocha.jl) - deep learning framework that Boltzmann.jl can export models to. 

Changing logistic function to tanh makes a little sense in terms of RBM, because result of logistic function is always between 0 and 1 and is interpreted as probability, while tanh can return negative values. But in general, if you want to introduce some new computation, you should create new RBM type inherited from abstract RBM (i.e. `type SomeRBM <: RBM ... end`) and override needed functions for it.  If you have concrete requirements, feel free to submit issue/feature request. 

Let me know if this answers your questions and/or you have any others.

**Comment by dfdx** [comment]
@sanjanasri: do you have any more questions or I can close this issue?

**Comment by sanjanasri** [comment]
Hi,

   I first apologies for replying you late. Thanks a lot for your earnest reply.  I saw the mocha export example. I can understand, but not clearly. Please let me know what i have understood is right. So, let me tell you what I have understood.

The parameters ( that is weight,bias etc) to a temporary  file. 
You give specified MemoryDataLayer, where you store original data and labels, then comes InnerProductLayer, where you define the structure similar to the one designed in dbn; Is Softmax layer specified is used later  to  perform classification for the deep learned data and calculates the loss comparing with MemoryDataLayer rite.

 what does "net = Net("TEST", backend, [data, vis, hid1, hid2])" do. If i don’t perform classification and only use the dbn only to learn the features;Can I ommit data and softmax layer.  Can it be rewritten as "net = Net("TEST", backend, [vis, hid1, hid2])" . 

Then at last, the parameters saved in temporary hdf5 file are loaded to the net.  So does this mean it is not necessary to perform the following in Mocha 
"params = SolverParameters(max_iter=10000, regu_coef=0.0005,
    mom_policy=MomPolicy.Fixed(0.9),
    lr_policy=LRPolicy.Inv(0.01, 0.0001, 0.75),
    load_from=exp_dir)
solver = SGD(params)"

And also,

If at all I wish to see the weights of each layer how can I get it in terms of net.

**Comment by dfdx** [comment]
If you are not going to build classifier, but only need to pretrain network with DBN and get resulting weight matrices for each layer, there's no need to use Mocha.jl at all - just train DBN:

```
using Boltzmann

X = # your dataset 

# specify layers of deep belief networks
layers = [("vis", GRBM(784, 256)),
               ("hid1", BernoulliRBM(256, 100)),
               ("hid2", BernoulliRBM(100, 100))]

# create DBN, which simply holds layers and provide convenient functions to work will of them at once
dbn = DBN(layers)

# train DBN with your data
fit(dbn, X)

# get n-th layer, i.e. n-th RBM
n = 1
rbm_name, rbm = dbn.layers[n]   # should be ("vis", GRBM(784, 256)) in this example

# get weight matrix for this layer
weights = components(rbm)
```

Can I ask you what you are trying to achieve? Maybe I can provide better instructions knowing your specific task.

**Comment by sanjanasri** [comment]
Thank you for ur reply. Actually I am planning to develop two architecture 
1)  Pre training DBN . That I completed.
 Following is the architecture
![image fqhk2x](https://cloud.githubusercontent.com/assets/13214774/8977256/1143f61e-36b5-11e5-94c8-7c9cfe74f0e5.png)

2) Developing DAE (Deep Autoencoder). After the unsupervised pre-training, DBNs are “unrolled” to create a semi-supervised DAE, which is then fine-tuned using back-propagation of error derivatives.
![image 5awb2x](https://cloud.githubusercontent.com/assets/13214774/8977268/19949738-36b5-11e5-9aa1-84037a50e47c.png)

In previous post, U mentioned that to unfold the network. I need to use mocha. That is where, I am unable to understand perfectly how should I export the model to mocha. 
And in Mocha, how can I see the weights and unfold them. I am sorry, If I am troubling you a lot

**Comment by dfdx** [comment]
Now I see your point. Physically it's not too hard, but there's no convenient interface for it right now. 

How urgent do you need it?

**Comment by sanjanasri** [comment]
May be in one or two weeks!!!..

**Comment by dfdx** [comment]
Seems like the easiest way would be to provide support for DAE here, in Boltzmann.jl, but leave all supervised learning/tuning to Mocha.jl. So what you want should be achievable something like this: 

```
X = ...
dbn = ...
fit(dbn, X)
dae = unroll(dbn)         # object of new type DAE will be created here
save_params("/path/to/file.hdf5", dae)
```
Saved parameters will be identical in format to normal DBN, so you will be able to load them into Mocha's network of appropriate structure and fine-tune it with its backpropagation algorithm. 

It's not hard to implement, so I'll do it today or tomorrow. However, I don't have much time to test it. Does it make sense for you to test it on your side with your specific use case? 

Also, I more or less understand architecture of Mocha's, but it's still better to ask the author what is the proper way to get weights from a trained network.

**Comment by dfdx** [comment]
You can check DAE support in [deep-autoencoder](https://github.com/dfdx/Boltzmann.jl/tree/deep-autoencoder) branch. Once tested, I will merge these changes to master. For now, you can `git checkout` to this branch and run example above.

**Comment by sanjanasri** [comment]
thanks a lot.... Here, I got some problem in getting internet connection.  I will just check and let u knw. If I have any doubts within today

**Comment by sanjanasri** [comment]
why am i getting an error "unroll not defined" even aftr updating.

**Comment by dfdx** [comment]
Did you do

    git checkout deep-autoencoder

from your `~/.julia/v0.3/Boltzmann` directory?

**Comment by sanjanasri** [comment]
when i tried doing git checkout I got an error "error: pathspec 'deep-autoencoder' did not match any file(s) known to git."

**Comment by dfdx** [comment]
Try 

    git fetch 
    git checkout deep-autoencoder

from the same directory. If it doesn't help, enter 

    git status
    
and post response here, please.

**Comment by sanjanasri** [comment]
![screenshot from 2015-08-05 12 59 52](https://cloud.githubusercontent.com/assets/13214774/9080308/059f5ae6-3b70-11e5-8808-d535d77c9b0d.png)

**Comment by sanjanasri** [comment]
That was the error I am getting

**Comment by dfdx** [comment]
You modified code in `dbn.jl`, thus you cannot checkout another branch. If you modified it occasionally and don't need these changes, dismiss them using:

    git reset --hard HEAD

if you need them, save them into a different branch:

    git checkout -b my-new-branch
    git add .
    git commit -m 'my changes'

and finally switch to the `deep-autoencoder` branch

    git checkout deep-autoencoder

Make sure every operation succeeded - these operations should be applied one after another, so if one command failed, the rest will not take affect too.

**Comment by sanjanasri** [comment]
I did something like below. screen shot
![screenshot from 2015-08-05 15 48 07](https://cloud.githubusercontent.com/assets/13214774/9083377/84a39ffc-3b87-11e5-9ca6-85a8df69e63e.png)
 When I tried for third comment i got an error 
```
*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'sanjana@sanjana-HP-ProDesk-400-G2-MT.(none)')
```

So i did something like this ```git config --global user.email "sanjanashree@gmail.com"```
then executed the third step 
```
On branch my-new-branch
nothing to commit, working directory clean
```

**Comment by sanjanasri** [comment]
I am getting an error while saving parameters to hdf5 file 
```
Error evaluating /home/sanjana/Documents/Julia_Pgms/dae.jl:
Error creating dataset //hid2_inv___weight
 in error at error.jl:22
 in d_create at /home/sanjana/.julia/v0.3/HDF5/src/plain.jl:747
 in _d_create at /home/sanjana/.julia/v0.3/HDF5/src/plain.jl:1467
 in _d_write at /home/sanjana/.julia/v0.3/HDF5/src/plain.jl:1500
 in write at /home/sanjana/.julia/v0.3/HDF5/src/plain.jl:1536
 in save_params at /home/sanjana/.julia/v0.3/Boltzmann/src/io.jl:19
 in anonymous at /home/sanjana/.julia/v0.3/Boltzmann/src/io.jl:23
 in h5open at /home/sanjana/.julia/v0.3/HDF5/src/plain.jl:568
 in save_params at /home/sanjana/.julia/v0.3/Boltzmann/src/io.jl:22
 in include_string at loading.jl:97
 in include_string at /home/sanjana/.julia/v0.3/Jewel/src/eval.jl:36
 in anonymous at /home/sanjana/.julia/v0.3/Jewel/src/LightTable/eval.jl:68
 in handlecmd at /home/sanjana/.julia/v0.3/Jewel/src/LightTable/LightTable.jl:65
 in handlenext at /home/sanjana/.julia/v0.3/Jewel/src/LightTable/LightTable.jl:81
 in server at /home/sanjana/.julia/v0.3/Jewel/src/LightTable/LightTable.jl:22
 in server at /home/sanjana/.julia/v0.3/Jewel/src/Jewel.jl:15
 in include at ./boot.jl:245
 in include_from_node1 at loading.jl:128
 in process_options at ./client.jl:285
 in _start at ./client.jl:354
```

and so the hdf5 seems to be incomplete.
![screenshot from 2015-08-05 17 35 49](https://cloud.githubusercontent.com/assets/13214774/9084984/8a76d930-3b96-11e5-9f9d-6437deed85fd.png)

**Comment by dfdx** [comment]
Fixed, type:

    git pull

and try again.

**Comment by sanjanasri** [comment]
Thanx a lot..... itz a gr8 work from u...

**Comment by dfdx** [comment]
Glad to hear that it helped you. Do you need any other help or I can merge new branch and close the issue?

**Comment by sanjanasri** [comment]
Yeah sure, this issue can be closed. Thanks again.
